File: tui/internal/ui/theme/theme.go
// Package theme holds the palette for the hibiscus TUI.
//
// The base surfaces + role-marker colors are Catppuccin Mocha (matching
// cli/src/ui/theme.rs) so chat content looks identical between the two
// clients. The brand accent is **hibiscus pink** — a vibrant tropical
// pink-red (#ff5d8f) that gives the TUI its own identity separate from
// the cli's peach accent. Three hibiscus tones:
//
//   Hibiscus      #ff5d8f  — primary accent (borders, badges, wordmark)
//   HibiscusGlow  #ff8fb1  — lighter, for "glow" emphasis (active dots)
//   HibiscusDim   #c43658  — darker, for backgrounds / pressed states
package theme

import "github.com/charmbracelet/lipgloss"

// Base palette — Catppuccin Mocha surfaces + role markers.
// Values mirror cli/src/ui/theme.rs color-for-color so chat content
// looks identical between the two clients.
var (
	// Hibiscus brand accent — the TUI's identity color. Used for
	// borders on the input box + connect modal, the "● connected"
	// status dot, the wordmark in the header, and the streaming
	// indicator. Three tones for visual hierarchy.
	Hibiscus     = lipgloss.Color("#ff5d8f") // primary — vibrant tropical pink
	HibiscusGlow = lipgloss.Color("#ff8fb1") // lighter — for glow / emphasis
	HibiscusDim  = lipgloss.Color("#c43658") // darker — for backgrounds, "pressed"

	AccentAlt = lipgloss.Color("#cba6f7") // mauve (secondary accent)

	// Role markers — kept identical to the cli so chat content reads
	// the same in both clients. These are semantic (User=green,
	// Assistant=sky) and intentionally NOT hibiscus-tinted.
	User      = lipgloss.Color("#a6e3a1") // green
	Assistant = lipgloss.Color("#89dceb") // sky
	System    = lipgloss.Color("#b4befe") // lavender
	Tool      = lipgloss.Color("#89dceb") // sky
	ToolError = lipgloss.Color("#f38ba8") // red

	// Surfaces
	FG       = lipgloss.Color("#cdd6f4") // text
	FGDim    = lipgloss.Color("#6c7086") // overlay0
	FGDimmer = lipgloss.Color("#454759") // surface1

	BGBase = lipgloss.Color("#1c1c1c") // neutral dark grey
	BGChat = lipgloss.Color("#3c3c3c") // elevated input surface

	BGCard       = lipgloss.Color("#313244") // catppuccin surface0
	BGCardHover  = lipgloss.Color("#585b70") // catppuccin surface2
	BGInspector  = lipgloss.Color("#23243a")
	BGInspectorH = lipgloss.Color("#3c3e54")

	// Borders — hibiscus when active (matches brand), dim otherwise.
	BorderActive = Hibiscus
	BorderIdle   = FGDimmer

	// Status / diff
	Success = User
	Error   = ToolError
	Warning = lipgloss.Color("#f9e2af") // yellow

	DiffAdd = lipgloss.Color("#a6e3a1")
	DiffRm  = lipgloss.Color("#f38ba8")

	// Semantic state colors. Tooling/Thinking now use the hibiscus
	// brand accent so the "agent is doing something" state matches
	// the rest of the TUI's identity.
	StateIdle       = FGDim
	StateThinking   = AccentAlt
	StateTooling    = Hibiscus
	StateCompacting = lipgloss.Color("#89b4fa")

	PermissionInfo        = Warning
	PermissionDestructive = ToolError

	// Tab + todo
	TabActive   = Hibiscus
	TabInactive = FGDim
	TodoPending = lipgloss.Color("#a6adc8")
	TodoActive  = Hibiscus
	TodoDone    = User
)

// RoleLabelStyle returns the lipgloss style for a chat message's
// role label. Mirrors cli/src/ui/theme.rs::role_label — role colors
// are NOT hibiscus-tinted, they stay semantic so chat content reads
// the same in both clients.
func RoleLabelStyle(role string) lipgloss.Style {
	switch role {
	case "user":
		return lipgloss.NewStyle().Foreground(User).Bold(true)
	case "assistant":
		return lipgloss.NewStyle().Foreground(Hibiscus).Bold(true)
	case "info", "summary":
		return lipgloss.NewStyle().Foreground(System).Bold(true)
	case "tool":
		return lipgloss.NewStyle().Foreground(Tool).Bold(true)
	default:
		return lipgloss.NewStyle().Foreground(FGDim).Bold(true)
	}
}

// GutterStyle is the colored vertical bar that sits under each
// speaker label.
func GutterStyle(role string) lipgloss.Style {
	c := FGDimmer
	switch role {
	case "user":
		c = User
	case "assistant":
		c = Assistant
	case "info", "summary":
		c = System
	case "tool":
		c = Tool
	}
	return lipgloss.NewStyle().Foreground(c)
}

File: src/ui/theme.rs
//! Central theme — Catppuccin Mocha-derived palette + helper builders.
//!
//! Every renderer pulls colors and `Block` builders from here instead of
//! inlining `Color::Cyan` / `Color::Magenta` ad-hoc. That keeps the look
//! coherent across header, sidebar, chat, popups, and viewer.

use ratatui::{
    style::{Color, Modifier, Style},
    text::Span,
    widgets::{Block, BorderType, Borders},
};

/// Color tokens. Values come from Catppuccin Mocha; the palette is
/// documented at <https://catppuccin.com/palette>. RGB literals render
/// on truecolor terminals; ratatui's terminal layer degrades to
/// 256-color or 16-color on legacy terminals automatically.
pub mod color {
    use ratatui::style::Color;

    // Accents
    pub const ACCENT: Color = Color::Rgb(250, 179, 135); // peach
    #[allow(dead_code)] // exposed for future breath-color / focus gradient use
    pub const ACCENT_DIM: Color = Color::Rgb(180, 130, 95);
    pub const ACCENT_ALT: Color = Color::Rgb(203, 166, 247); // mauve

    // Role markers
    pub const USER: Color = Color::Rgb(166, 227, 161); // green
    pub const ASSISTANT: Color = Color::Rgb(137, 220, 235); // sky
    pub const SYSTEM: Color = Color::Rgb(180, 190, 254); // lavender
    pub const TOOL: Color = Color::Rgb(137, 220, 235); // sky
    pub const TOOL_ERROR: Color = Color::Rgb(243, 139, 168); // red

    // Surfaces
    pub const FG: Color = Color::Rgb(205, 214, 244); // text
    pub const FG_DIM: Color = Color::Rgb(108, 112, 134); // overlay0
    pub const FG_DIMMER: Color = Color::Rgb(69, 71, 90); // surface1

    // ── OpenCode-style flat surface palette ──────────────────────────
    // Neutral greys — no blue tint. Chat area + sidebar + inspector
    // share the deepest dark grey (BG_BASE); the input box is the
    // only lighter-grey surface (BG_CHAT) so it reads as the focused
    // affordance without a border.
    //
    //   BG_BASE  (28, 28, 28)  — chat, sidebar, inspector, chrome
    //   BG_CHAT  (60, 60, 60)  — input box (the only elevated surface)
    /// Deepest layer — used for the global terminal bg + chat +
    /// sidebar + inspector + header + status bar. Neutral dark grey,
    /// no blue tint.
    pub const BG_BASE: Color = Color::Rgb(28, 28, 28);
    /// Elevated surface — used for the input box only. Light grey
    /// one step up from BG_BASE so the box reads as the focused
    /// affordance without a border.
    pub const BG_CHAT: Color = Color::Rgb(60, 60, 60);

    // Borders
    pub const BORDER_ACTIVE: Color = ACCENT; // peach
    pub const BORDER_IDLE: Color = Color::Rgb(69, 71, 90); // surface1

    // Card surface — used to group consecutive read-only tool messages into
    // a single borderless tile (see `ui::chat::render_chat`). One step up
    // from the chat panel's transparent canvas so the card reads as a
    // distinct, slightly-elevated block without needing a frame.
    pub const BG_CARD: Color = Color::Rgb(49, 50, 68); // catppuccin surface0
    /// Hovered card row — one elevation brighter than `BG_CARD` so the
    /// pointer's current target reads clearly as "clickable" without a
    /// chevron or arrow icon.
    pub const BG_CARD_HOVER: Color = Color::Rgb(88, 91, 112); // catppuccin surface2

    // Status / diff
    pub const SUCCESS: Color = USER;
    pub const ERROR: Color = TOOL_ERROR;
    pub const WARNING: Color = Color::Rgb(249, 226, 175); // yellow

    // ── Semantic palette (agentic UI) ─────────────────────────────────
    // Colors that mean something *in context* — diffs, permissions,
    // todos, agent state. Re-using base Catppuccin tokens where possible
    // so the new vocabulary stays cohesive with the existing look.

    /// Diff additions — green that reads clearly on both the chat surface
    /// (dark) and the BG_CARD inspector tile. Same hue as USER but kept
    /// as its own constant so a future theme tweak (e.g. dim green for
    /// "added by tool, not by user") has one knob.
    pub const DIFF_ADD: Color = Color::Rgb(166, 227, 161); // green
    /// Diff removals — desaturated red so it doesn't shout on a card.
    pub const DIFF_REMOVE: Color = Color::Rgb(243, 139, 168); // red

    /// A pending todo item — neutral dim, indicating "not started yet".
    pub const TODO_PENDING: Color = Color::Rgb(166, 173, 200); // subtext1
    /// A todo currently in progress — peach (the agent's accent), so it
    /// visually matches the active-tool dot in the inspector.
    pub const TODO_ACTIVE: Color = Color::Rgb(250, 179, 135); // peach
    /// A completed todo — green, matching the user/assistant-success palette.
    pub const TODO_DONE: Color = Color::Rgb(166, 227, 161); // green

    /// Read-only permission prompt (e.g. share a file path back to the
    /// user). Soft yellow — "look at this", not "watch out".
    pub const PERMISSION_INFO: Color = Color::Rgb(249, 226, 175); // yellow
    /// Destructive action — `run_command`, `write_file` overwrite,
    /// `forget_memory`. Bright red so the card edges scream for attention.
    pub const PERMISSION_DESTRUCTIVE: Color = Color::Rgb(243, 139, 168); // red
    /// Modifier-style destructive background (low-alpha red on top of the
    /// panel bg). Use as `bg(...)` only — not as fg, would clash.
    pub const PERMISSION_DESTRUCTIVE_BG: Color = Color::Rgb(70, 40, 50); // muted maroon

    /// Agent state badge colors — Idle (dim), Thinking (mauve), Tooling
    /// (peach). Inspector header uses these for the small "● THINKING"
    /// chip in the top-right.
    pub const STATE_IDLE: Color = Color::Rgb(108, 112, 134); // overlay0
    pub const STATE_THINKING: Color = Color::Rgb(203, 166, 247); // mauve
    pub const STATE_TOOLING: Color = Color::Rgb(250, 179, 135); // peach
    pub const STATE_COMPACTING: Color = Color::Rgb(137, 180, 250); // blue

    /// Inspector background — one step deeper than BG_CARD so the right
    /// column reads as the "system/dashboard" surface vs. the chat surface
    /// (which is transparent over the base bg).
    pub const BG_INSPECTOR: Color = Color::Rgb(35, 36, 50); // between base & surface0
    /// Hovered inspector row — matches BG_CARD_HOVER for parity with chat.
    pub const BG_INSPECTOR_HOVER: Color = Color::Rgb(60, 62, 84);

    /// Sidebar selected-tab indicator (active sessions/agents/files tab).
    pub const TAB_ACTIVE: Color = Color::Rgb(250, 179, 135); // peach
    /// Sidebar inactive tab text.
    pub const TAB_INACTIVE: Color = Color::Rgb(108, 112, 134); // overlay0
}

/// Horizontal inset inside any panel border. One col on each side gives
/// content breathing room without burning real estate on narrow terms.
#[allow(dead_code)] // documents the convention; renderers currently call Padding::horizontal(1) directly
pub const PANEL_PAD_H: u16 = 1;

/// Build a rounded panel-style `Block` with focus-aware border colour.
/// Pass an empty string for `title` to render a borderless-title rounded
/// container.
pub fn panel_block(title: &str, focused: bool) -> Block<'_> {
    let border_color = if focused {
        color::BORDER_ACTIVE
    } else {
        color::BORDER_IDLE
    };
    let mut block = Block::default()
        .borders(Borders::ALL)
        .border_type(BorderType::Rounded)
        .border_style(Style::default().fg(border_color));
    if !title.is_empty() {
        block = block.title(Span::styled(
            format!(" {title} "),
            Style::default()
                .fg(if focused {
                    color::ACCENT
                } else {
                    color::FG_DIM
                })
                .add_modifier(Modifier::BOLD),
        ));
    }
    block
}

/// Build a rounded "popup" block — distinctly styled with the alt-accent
/// (mauve) so popups feel like overlays rather than another panel.
/// `danger=true` swaps the border to red for destructive confirmations
/// (run_command, write_file with diff that drops content, etc.).
pub fn popup_block(title: &str, danger: bool) -> Block<'_> {
    let border_color = if danger {
        color::ERROR
    } else {
        color::ACCENT_ALT
    };
    Block::default()
        .borders(Borders::ALL)
        .border_type(BorderType::Rounded)
        .border_style(Style::default().fg(border_color))
        .title(Span::styled(
            format!(" {title} "),
            Style::default()
                .fg(border_color)
                .add_modifier(Modifier::BOLD),
        ))
}

/// Role label + color used at the head of each chat message.
/// The leading `▎` is a colored vertical bar that continues down the
/// message body too (see `chat::render_chat`), giving each turn an
/// OpenCode-style "card with a colored gutter" look.
pub fn role_label(role: &str) -> (&'static str, Color) {
    match role {
        "user" => ("▎ user", color::USER),
        "assistant" => ("▎ assistant", color::ASSISTANT),
        "info" => ("▎ system", color::SYSTEM),
        "summary" => ("▎ compacted", color::SYSTEM),
        "tool" => ("▎ tool", color::TOOL),
        _ => ("?", color::FG_DIM),
    }
}
